pytorch_lightning/utilities/device_parser.py

```diff
@@ -120,11 +120,13 @@ def parse_tpu_cores(tpu_cores: Union[int, str, List]) -> Optional[Union[List[int
 
     if not _tpu_cores_valid(tpu_cores):
         raise MisconfigurationException("`tpu_cores` can only be 1, 8 or [<1-8>]")
+    else:  # need to inform mypy `tpu_cores` has now a right type
+        tpu_cores_parsed: Optional[Union[int, List[int]]] = tpu_cores
```

Comment:
I believe you can use `TypeGuard` from `typing_extensions` which we already install:

https://github.com/PyTorchLightning/pytorch-lightning/blob/2e537b75e3ac50e0218cf040e66d04cb496149f2/requirements.txt#L13

More info:

https://github.com/python/typing/blob/master/typing_extensions/src_py3/typing_extensions.py#L2561

https://www.python.org/dev/peps/pep-0647/

Reply:
Thanks a lot for this suggestion, I'm gonna take a look!